libnetdata/locks/locks.c

```diff
@@ -171,173 +211,450 @@ int __netdata_rwlock_init(netdata_rwlock_t *rwlock) {
     int ret = __netdata_rwlock_rdlock(rwlock);
+    usec_t end_s = now_monotonic_high_precision_usec();
+
+    if(!ret) {
+        locker = update_or_add_rwlock_locker(file, function, line, rwlock, locker, 'R');
+        if(log) log_rwlock_lockers(file, function, line, rwlock, "GOT", 'R');
 
-    debug(D_LOCKS, "RW_LOCK: netdata_rwlock_rdlock(0x%p) = %d in %llu usec, from %lu@%s, %s()", rwlock, ret, now_boottime_usec() - start, line, file, function);
+    }
+
+    if(end_s - start_s > NETDATA_TRACE_RWLOCKS_WAIT_TIME_TO_IGNORE_USEC)
+        fprintf(stderr,
+                "RW_LOCK ON LOCK 0x%p: %d, '%s' (function %s() %lu@%s) WAITED for a READ lock for %llu usec.\n",
+                rwlock,
+                gettid(), netdata_thread_tag(),
+                function, line, file,
+                end_s - start_s);
+
+    debug(D_LOCKS, "RW_LOCK: netdata_rwlock_rdlock(0x%p) = %d in %llu usec, from %lu@%s, %s()", rwlock, ret, end_s - start_s, line, file, function);
 
     return ret;
 }
 
 int netdata_rwlock_wrlock_debug(const char *file __maybe_unused, const char *function __maybe_unused,
                                 const unsigned long line __maybe_unused, netdata_rwlock_t *rwlock) {
-    usec_t start = 0;
-    (void)start;
 
-    if(unlikely(debug_flags & D_LOCKS)) {
-        start = now_boottime_usec();
-        debug(D_LOCKS, "RW_LOCK: netdata_rwlock_wrlock(0x%p) from %lu@%s, %s()", rwlock, line, file, function);
+    debug(D_LOCKS, "RW_LOCK: netdata_rwlock_wrlock(0x%p) from %lu@%s, %s()", rwlock, line, file, function);
+
+    netdata_rwlock_locker *locker = find_rwlock_locker(file, function, line, rwlock);
+    if(locker)
+        not_supported_by_posix_rwlocks(file, function, line, rwlock, 'W', "DEADLOCK - WANTS A WRITE LOCK BUT ALREADY HAVE THIS LOCKED");
+
+    int log = 0;
+    if(rwlock->readers) {
+        log_rwlock_lockers(file, function, line, rwlock, "WANTS", 'W');
+        log = 1;
     }
 
+    usec_t start_s = now_monotonic_high_precision_usec();
     int ret = __netdata_rwlock_wrlock(rwlock);
+    usec_t end_s = now_monotonic_high_precision_usec();
+
+    if(!ret){
+        locker = update_or_add_rwlock_locker(file, function, line, rwlock, locker, 'W');
+        if(log) log_rwlock_lockers(file, function, line, rwlock, "GOT", 'W');
+    }
 
-    debug(D_LOCKS, "RW_LOCK: netdata_rwlock_wrlock(0x%p) = %d in %llu usec, from %lu@%s, %s()", rwlock, ret, now_boottime_usec() - start, line, file, function);
+    if(end_s - start_s > NETDATA_TRACE_RWLOCKS_WAIT_TIME_TO_IGNORE_USEC)
+        fprintf(stderr,
+                "RW_LOCK ON LOCK 0x%p: %d, '%s' (function %s() %lu@%s) WAITED for a WRITE lock for %llu usec.\n",
+                rwlock,
+                gettid(), netdata_thread_tag(),
+                function, line, file,
+                end_s - start_s);
+
+    debug(D_LOCKS, "RW_LOCK: netdata_rwlock_wrlock(0x%p) = %d in %llu usec, from %lu@%s, %s()", rwlock, ret, now_monotonic_usec() - start, line, file, function);
```

Comment:
Compilation error on variable `start`.
3rd arg `now_monotonic_usec() - start` should be `now_monotonic_usec() - start_s`

Reply:
fixed.